File: testbed_ruby/test/associations/n_to_optional_n_test.rb
require 'test/unit'
require './src-gen-umple/student_ac'
require './src-gen-umple/mentor_ac'
require './src-gen-umple/program_ac'

module CruiseAssociations
class NToOptionalNTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentors
    m = MentorAC.new("blah")
    StudentAC.new(99)
  end
  
  def test_CreateStudentJustEnoughMentors
    m = MentorAC.new("blah")
    m2 = MentorAC.new("blah2")
    m3 = MentorAC.new("blah2")
    s = StudentAC.new(99)
    
    assert_equal(true,s.add_mentor(m))
    assert_equal(true,s.add_mentor(m2))
    assert_equal(false,s.add_mentor(m3))
    assert_equal(2,s.number_of_mentors)
  end

  def test_CreateMentorWithoutStudent
    m = MentorAC.new("blah")
    assert_equal(0,m.number_of_students)
  end
  
  def test_SetMentorsOutsideRange
    s = StudentAC.new(99)

    m = MentorAC.new("blah")
    m2 = MentorAC.new("blah")
    m3 = MentorAC.new("blah")

    assert_equal(true,s.add_mentor(m))
    assert_equal(true,s.add_mentor(m2))
    assert_equal(false,s.add_mentor(m3))
    
    assert_equal(2, s.number_of_mentors)
    assert_equal(s,m.get_student(0))
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m3.number_of_students)
  end
  
  def test_AddStudents
    m = MentorAC.new("blah")
    
    s = StudentAC.new(99)
    s2 = StudentAC.new(98)
    s3 = StudentAC.new(97)
    s4 = StudentAC.new(96)

    assert_equal(true,m.add_student(s))
    assert_equal(true,m.add_student(s2))
    assert_equal(true,m.add_student(s3))
    assert_equal(false,m.add_student(s4))
    
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(1,s3.number_of_mentors)
  end   
  
  def test_RemoveMiddleMentorWhenNotValidMaintainsTheOrder
    s = StudentAC.new(99)
    
    m = MentorAC.new("blah")
    m2 = MentorAC.new("blah")

    s.add_mentor(m)
    s.add_mentor(m2)
    assert_equal(false,s.remove_mentor(m2))
    
    assert_equal(2, s.number_of_mentors)
    assert_equal(m2,s.get_mentor(1))
  end 
  
  def test_CannotRemoveStudents
    m = MentorAC.new("blah")
    m2 = MentorAC.new("blah2")
    m3 = MentorAC.new("blah3")
    
    s = StudentAC.new(99)
    s2 = StudentAC.new(98)
    s3 = StudentAC.new(97)
    
    m.add_student(s)
    m.add_student(s2)
    m2.add_student(s)
    m2.add_student(s2)
    assert_equal(false,m3.add_student(s))
    assert_equal(false,m3.add_student(s2))
    
    assert_equal(false,m.remove_student(s3))
    assert_equal(false,m.remove_student(s2))
    
    assert_equal(2,s2.number_of_mentors)
    assert_equal(2,m.number_of_students)
  end 
  
  def test_SetStudentsTooMany
    m = MentorAC.new("blah")
    
    s = StudentAC.new(99)
    s2 = StudentAC.new(98)
    s3 = StudentAC.new(97)
    s4 = StudentAC.new(96)
    
    m.add_student(s)
    m.add_student(s2)
    assert_equal(true,m.add_student(s3))
    assert_equal(false,m.add_student(s4))
  end   
  
  def test_MentorAlreadyHasEnoughStudents
    m = MentorAC.new("blah")
    
    s = StudentAC.new(99)
    s2 = StudentAC.new(98)
    s3 = StudentAC.new(97)
    s4 = StudentAC.new(96)
    
    m.add_student(s)
    m.add_student(s2)
    m.add_student(s3)
    
    assert_equal(false,m.add_student(s4))
    
    assert_equal(3,m.number_of_students)
    assert_equal(1,s3.number_of_mentors)
  end    

  def test_add_mentor
    m = MentorAC.new("blah")
    m2 = MentorAC.new("blah2")
    m3 = MentorAC.new("blah2")

    s = StudentAC.new(99)
    s2 = StudentAC.new(98)
    
    assert_equal(true,m.add_student(s))
    assert_equal(true,m.add_student(s2))
    assert_equal(true,s.add_mentor(m2))
    assert_equal(false,s.add_mentor(m3))

    assert_equal(2,m.number_of_students)
    assert_equal(1,m2.number_of_students)
    assert_equal(0,m3.number_of_students)
    assert_equal(2,s.number_of_mentors)
  end 
  
  def test_getBoundsForStudent
    assert_equal(3,MentorAC::maximum_number_of_students)
  end  
  
  def test_getBoundsForMentor
    assert_equal(2,StudentAC::minimum_number_of_mentors)
    assert_equal(2,StudentAC::maximum_number_of_mentors)
  end    
  

  def test_deleteMentorAndStudentNowHasTooFewMentors
    m = MentorAC.new("blah")
    m2 = MentorAC.new("blah2")
    m3 = MentorAC.new("blah2")

    s = StudentAC.new(99)
    s2 = StudentAC.new(98)
    
    s.add_mentor(m)
    s.add_mentor(m2)
    s.add_mentor(m3)
    s2.add_mentor(m)
    s2.add_mentor(m2)
    s2.add_mentor(m3)
    
    m3.add_student(s)
    m3.add_student(s2)
    
    p = ProgramAC.new
    s.set_program(p)
    
    m.delete
    assert_equal(0,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    assert_equal(0,s.number_of_mentors)
  end  

end
end

File: testbed_php/test/associations/NToOptionalNTest.php
<?php

class NToOptionalNTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentors()
  {
    $m = new MentorAC("blah");
    new StudentAC(99);
  }
  

  public function test_CreateStudentJustEnoughMentors()
  {
    $m = new MentorAC("blah");
    $m2 = new MentorAC("blah2");
    $m3 = new MentorAC("blah2");
    $s = new StudentAC(99);
    
    $this->assertEqual(true,$s->addMentor($m));
    $this->assertEqual(true,$s->addMentor($m2));
    $this->assertEqual(false,$s->addMentor($m3));
    $this->assertEqual(2,$s->numberOfMentors());
  }


  public function test_CreateMentorWithoutStudent()
  {
    $m = new MentorAC("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }
  

  public function test_SetMentorsOutsideRange()
  {
    $s = new StudentAC(99);

    $m = new MentorAC("blah");
    $m2 = new MentorAC("blah");
    $m3 = new MentorAC("blah");

    $this->assertEqual(true,$s->addMentor($m));
    $this->assertEqual(true,$s->addMentor($m2));
    $this->assertEqual(false,$s->addMentor($m3));
    
    $this->assertEqual(2, $s->numberOfMentors());
    $this->assertEqual($s,$m->getStudent(0));
    $this->assertEqual($s,$m2->getStudent(0));
    $this->assertEqual(0,$m3->numberOfStudents());
  }
  

  public function test_AddStudents()
  {
    $m = new MentorAC("blah");
    
    $s = new StudentAC(99);
    $s2 = new StudentAC(98);
    $s3 = new StudentAC(97);
    $s4 = new StudentAC(96);

    $this->assertEqual(true,$m->addStudent($s));
    $this->assertEqual(true,$m->addStudent($s2));
    $this->assertEqual(true,$m->addStudent($s3));
    $this->assertEqual(false,$m->addStudent($s4));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual(1,$s3->numberOfMentors());
  }   
  

  public function test_RemoveMiddleMentorWhenNotValidMaintainsTheOrder()
  {
    $s = new StudentAC(99);
    
    $m = new MentorAC("blah");
    $m2 = new MentorAC("blah");

    $s->addMentor($m);
    $s->addMentor($m2);
    $this->assertEqual(false,$s->removeMentor($m2));
    
    $this->assertEqual(2, $s->numberOfMentors());
    $this->assertEqual($m2,$s->getMentor(1));
  } 
  

  public function test_CannotRemoveStudents()
  {
    $m = new MentorAC("blah");
    $m2 = new MentorAC("blah2");
    $m3 = new MentorAC("blah3");
    
    $s = new StudentAC(99);
    $s2 = new StudentAC(98);
    $s3 = new StudentAC(97);
    
    $m->addStudent($s);
    $m->addStudent($s2);
    $m2->addStudent($s);
    $m2->addStudent($s2);
    $this->assertEqual(false,$m3->addStudent($s));
    $this->assertEqual(false,$m3->addStudent($s2));
    
    $this->assertEqual(false,$m->removeStudent($s3));
    $this->assertEqual(false,$m->removeStudent($s2));
    
    $this->assertEqual(2,$s2->numberOfMentors());
    $this->assertEqual(2,$m->numberOfStudents());
  } 
  

  public function test_SetStudentsTooMany()
  {
    $m = new MentorAC("blah");
    
    $s = new StudentAC(99);
    $s2 = new StudentAC(98);
    $s3 = new StudentAC(97);
    $s4 = new StudentAC(96);
    
    $m->addStudent($s);
    $m->addStudent($s2);
    $this->assertEqual(true,$m->addStudent($s3));
    $this->assertEqual(false,$m->addStudent($s4));
  }   
  

  public function test_MentorAlreadyHasEnoughStudents()
  {
    $m = new MentorAC("blah");
    
    $s = new StudentAC(99);
    $s2 = new StudentAC(98);
    $s3 = new StudentAC(97);
    $s4 = new StudentAC(96);
    
    $m->addStudent($s);
    $m->addStudent($s2);
    $m->addStudent($s3);
    
    $this->assertEqual(false,$m->addStudent($s4));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(1,$s3->numberOfMentors());
  }    


  public function test_addMentor()
  {
    $m = new MentorAC("blah");
    $m2 = new MentorAC("blah2");
    $m3 = new MentorAC("blah2");

    $s = new StudentAC(99);
    $s2 = new StudentAC(98);
    
    $this->assertEqual(true,$m->addStudent($s));
    $this->assertEqual(true,$m->addStudent($s2));
    $this->assertEqual(true,$s->addMentor($m2));
    $this->assertEqual(false,$s->addMentor($m3));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(1,$m2->numberOfStudents());
    $this->assertEqual(0,$m3->numberOfStudents());
    $this->assertEqual(2,$s->numberOfMentors());
  } 
  

  public function test_getBoundsForStudent()
  {
    $this->assertEqual(3,MentorAC::maximumNumberOfStudents());
  }  
  

  public function test_getBoundsForMentor()
  {
    $this->assertEqual(2,StudentAC::minimumNumberOfMentors());
    $this->assertEqual(2,StudentAC::maximumNumberOfMentors());
  }    
  

  public function test_deleteMentorAndStudentNowHasTooFewMentors()
  {
    
    $m = new MentorAC("blah");
    $m2 = new MentorAC("blah2");
    $m3 = new MentorAC("blah2");

    $s = new StudentAC(99);
    $s2 = new StudentAC(98);
    
    $s->addMentor($m);
    $s->addMentor($m2);
    $s->addMentor($m3);
    $s2->addMentor($m);
    $s2->addMentor($m2);
    $s2->addMentor($m3);
    
    $m3->addStudent($s);
    $m3->addStudent($s2);
    
    $p = new ProgramAC();
    $s->setProgram($p);
    
    $m->delete();
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    $this->assertEqual(0,$s->numberOfMentors());
  }  

}
